projects/media-viewer/src/lib/icp/icp.directive.ts

```diff
@@ -0,0 +1,97 @@
+import { Directive, Input, OnDestroy, OnInit } from '@angular/core';
+import { ToolbarEventService } from '../toolbar/toolbar-event.service';
+import { select, Store } from '@ngrx/store';
+import * as fromIcpSelectors from '../store/selectors/icp.selectors';
+import * as fromIcpActions from '../store/actions/icp.action';
+import { Subscription } from 'rxjs';
+import { IcpUpdateService } from './icp-update.service';
+import { ViewerEventService } from '../viewers/viewer-event.service';
+import { IcpPresenterService } from './icp-presenter.service';
+import { IcpFollowerService } from './icp-follower.service';
+import { filter, take } from 'rxjs/operators';
+import { IcpParticipant, IcpState } from './icp.interfaces';
+
+
+@Directive({
+  selector: '[inCourtPresentation]'
+})
+export class IcpDirective implements OnInit, OnDestroy {
+
+  @Input() caseId: string;
+
+  client: IcpParticipant;
+  presenter: IcpParticipant;
+  isPresenter: boolean;
+
+  subscription: Subscription;
+  sessionSubscription: Subscription;
+
+  constructor(private readonly toolbarEvents: ToolbarEventService,
+              private readonly viewerEvents: ViewerEventService,
+              private readonly socketService: IcpUpdateService,
+              private readonly presenterSubscriptions: IcpPresenterService,
+              private readonly followerSubscriptions: IcpFollowerService,
+              private store: Store<IcpState>) {}
+
+  ngOnInit() {
+    this.subscription = this.toolbarEvents.icp.sessionLaunch.subscribe(() => this.launchSession());
```

Comment:
icp.sessionLaunch is the step that can be removed and just dispatch action where u trigger .next()
if we have ngrx services should be really just for HTTP.

Reply:
not sure if I understand correctly, but we can't dispatch the action where we do trigger.next() because that happens in a toolbar, and toolbars are optional, so we can't put functionality in there. that's why trigger an event from the toolbar, and then handle it here, in the same way that someone using their own toolbar might do